src/cvi/modules/CH.py

```diff
@@ -1,97 +1,126 @@
-    # 3. M. Moshtaghi, J. C. Bezdek, S. M. Erfani, C. Leckie, and J. Bailey, "Online Cluster Validity Indices for Streaming Data," ArXiv e-prints, 2018, arXiv:1801.02937v1 [stat.ML]. [Online].
-    # 4. M. Moshtaghi, J. C. Bezdek, S. M. Erfani, C. Leckie, J. Bailey, "Online cluster validity indices for performance monitoring of streaming data clustering," Int. J. Intell. Syst., pp. 1-23, 2018.
-    # """
-    # Calinski-Harabasz (CH) Cluster Validity Index.
-    # """
-
     def __init__(self):
         """
         CH initialization routine.
         """
-        # """
-        # Test documentation.
-        # """
+
         super().__init__()
 
         return
-        # print("Hello world!")
-        # self.data = []
 
-    def param_inc(self, sample:np.ndarray, label:np.ndarray):
+    @add_docs(param_inc_doc)
+    def param_inc(self, sample:np.ndarray, label:int) -> None:
+        """
+        Incremental parameter update for the Calinski-Harabasz (CH) CVI.
+        """
+
+        # Get the internal label corresponding to the provided label
         i_label = self.label_map.get_internal_label(label)
 
+        # Increment the local number of samples count
         n_samples_new = self.n_samples + 1
+
+        # Check if the module has been setup, then set the mu accordingly
         if not self.mu.any():
             mu_new = sample
             self.setup(sample)
         else:
             mu_new = (1 - 1/n_samples_new) * self.mu + (1/n_samples_new) * sample
 
-        if i_label > self.n_clusters:
+        # IF NEW CLUSTER LABEL
+        # Correct for python 0-indexing
+        if i_label + 1 > self.n_clusters:
             n_new = 1
             v_new = sample
             CP_new = 0.0
             G_new = np.zeros(self.dim)
+
             # Update 1-D parameters with list appends
             self.n_clusters += 1
             self.n.append(n_new)
             self.CP.append(CP_new)
-            # Update 2-D parameters with numpy appends
-            self.v = np.append(self.v, v_new)
-            self.G = np.append(self.G, G_new)
+
+            # Update 2-D parameters with numpy vstacks
+            self.v = np.vstack([self.v, v_new])
+            self.G = np.vstack([self.G, G_new])
+
+        # ELSE OLD CLUSTER LABEL
         else:
             n_new = self.n[i_label] + 1
-            v_new = (1 - 1/n_new) * self.v[:, i_label] + (1/n_new) * sample
-            delta_v = self.v[:, i_label] - v_new
+            v_new = (1 - 1/n_new) * self.v[i_label, :] + (1/n_new) * sample
+            delta_v = self.v[i_label, :] - v_new
             diff_x_v = sample - v_new
-            CP_new = self.P[i_label] + transpose(diff_x_v[np.newaxis])
-            G_new = self.G[:, i_label] + diff_x_v + self.n[i_label] * delta_v
+            CP_new = self.CP[i_label] + np.inner(diff_x_v, diff_x_v) + self.n[i_label]*np.inner(delta_v, delta_v) + 2*np.inner(delta_v,self.G[i_label, :])
```

Comment:
What's goin on here? this looks computationally taxing

Reply:
This is the equivalent of your CP_update in icvi.py for CH. It has three vector dot products (np.inner is the same as np.dot for 1-D vectors, though I might change it to explicitly np.dot).